vault/resource_token.go

```diff
@@ -0,0 +1,301 @@
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The client token accessor.",
+			},
+			"client_token": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The client token.",
+				Sensitive:   true,
+			},
+		},
+	}
+}
+
+func tokenCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	role := d.Get("role_name").(string)
+
+	iPolicies := d.Get("policies").(*schema.Set).List()
+	policies := make([]string, 0, len(iPolicies))
+	for _, iPolicy := range iPolicies {
+		policies = append(policies, iPolicy.(string))
+	}
+
+	var createRequest = &api.TokenCreateRequest{}
+
+	if len(policies) > 0 {
+		createRequest.Policies = policies
+	}
+
+	if v, ok := d.GetOk("ttl"); ok {
+		createRequest.TTL = v.(string)
+	}
+
+	if v, ok := d.GetOk("explicit_max_ttl"); ok {
+		createRequest.ExplicitMaxTTL = v.(string)
+	}
+
+	if v, ok := d.GetOk("period"); ok {
+		createRequest.Period = v.(string)
+	}
+
+	if v, ok := d.GetOk("no_parent"); ok {
+		createRequest.NoParent = v.(bool)
+	}
+
+	if v, ok := d.GetOk("no_default_policy"); ok {
+		createRequest.NoDefaultPolicy = v.(bool)
+	}
+
+	if v, ok := d.GetOk("display_name"); ok {
+		createRequest.DisplayName = v.(string)
+	}
+
+	if v, ok := d.GetOk("num_uses"); ok {
+		createRequest.NumUses = v.(int)
+	}
+
+	if v, ok := d.GetOk("renewable"); ok {
+		renewable := v.(bool)
+		createRequest.Renewable = &renewable
+	}
+
+	var resp *api.Secret
+	var err error
+
+	if role != "" {
+		log.Printf("[DEBUG] Creating token with role %q", role)
+		resp, err = client.Auth().Token().CreateWithRole(createRequest, role)
+		if err != nil {
+			return fmt.Errorf("error creating token with role %q: %s", role, err)
+		}
+
+		log.Printf("[DEBUG] Created token %q with role %q", resp.Auth.Accessor, role)
+	} else {
+		log.Printf("[DEBUG] Creating token")
+		resp, err = client.Auth().Token().Create(createRequest)
+		if err != nil {
+			return fmt.Errorf("error creating token: %s", err)
+		}
+
+		log.Printf("[DEBUG] Created token %q", resp.Auth.Accessor)
+	}
+
+	d.Set("lease_duration", resp.Auth.LeaseDuration)
+	d.Set("lease_started", time.Now().Format(time.RFC3339))
+	d.Set("client_token", resp.Auth.ClientToken)
+
+	d.SetId(resp.Auth.Accessor)
+
+	return tokenRead(d, meta)
+}
+
+func tokenRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	id := d.Id()
+
+	log.Printf("[DEBUG] Reading token %q", id)
+	resp, err := client.Auth().Token().LookupAccessor(id)
+	if err != nil {
+		log.Printf("[WARN] Token not found, removing from state")
+		d.SetId("")
+		return nil
+	}
+
+	log.Printf("[DEBUG] Read token %q", id)
+
+	if tokenCheckLease(d, client) {
+		log.Printf("[DEBUG] Lease for %q expiring soon, renewing", d.Id())
+		renewed, err := client.Auth().Token().Renew(d.Get("client_token").(string), d.Get("lease_duration").(int))
+		if err != nil {
+			log.Printf("[DEBUG] Error renewing token %q, bailing", d.Id())
```

Comment:
Should this be returning here?

Reply:
Fixed, I clean the state to force a new resource creation.